src/tag_index.c

```diff
@@ -74,10 +70,28 @@ char **TagIndex_Preprocess(char sep, TagFieldFlags flags, const DocumentField *d
         tok = strtolower(tok);
       }
       tok = rm_strndup(tok, MIN(toklen, MAX_TAG_LEN));
-      ret = array_append(ret, tok);
+      *resArray = array_append(*resArray, tok);
     }
   }
   rm_free(pp);
+  return REDISMODULE_OK;
+}
+
+/* Preprocess a document tag field, returning a vector of all tags split from the content */
+char **TagIndex_Preprocess(char sep, TagFieldFlags flags, const DocumentField *data) {
+  char **ret = array_new(char *, 4);
+  if (data->unionType == FLD_VAR_T_RMS) {
+    const char *str = (char *)RedisModule_StringPtrLen(data->text, NULL);
+    tokenizeTagString(str, sep, flags, &ret);
+  } else if (data->unionType == FLD_VAR_T_CSTR) {
+    tokenizeTagString(data->strval, sep, flags, &ret);
+  } else if (data->unionType == FLD_VAR_T_ARRAY) {
+    for (int i = 0; i < data->arrayLen; i++) {
+      tokenizeTagString(data->multiVal[i], sep, flags, &ret);
```

Comment:
According to the spec, if separator was not specifically defined here then we should not tokenize, right @emmanuelkeller ?

Reply:
I actually think that we should not tokenize in this case at all, if the user gives an array we take the array. We can tokenize (if needed) when we read the json ...

Reply:
#2069 deals with it but it is based on the previous JSON API so it should be updated.
It can be added to this PR though I would prefer to separate it to keep the PR size down.

Reply:
So either we get a `list of tags` or an `array of tags` but we should not accept `an array of lists of tags`?

Reply:
I think we decided to tokenize only if TAG and only if a separator is explicitly defined (since in JSON there is a concept of array unlike in HASH, but we wanted to allow an existing string value to be still considered as multi if required)